compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -546,6 +546,29 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
 
 impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
     fn visit_operand(&mut self, operand: &Operand<'tcx>, location: Location) {
+        match operand {
+            Operand::Copy(place) | Operand::Move(place) => {
```

Comment:
Why is this only added in the `TypeChecker`? That one is only called by inlining. The main validator is further up this file, that describes the properties that we always expect all MIR to have.

If this is meant to be a new requirement for all MIR, however, then even the initially built MIR must satisfy it. It is very fragile to first have "invalid" MIR and then later patch it up.

(I don't know why we have two visitors, there are no comments that I could find that explain any of this.  But adding this check here won't do a lot I think.)

Reply:
Cc @cjgillot for the two-visitor situation -- which visitor should take new check?

Reply:
> Could the implementation be simplifies by moving the check itself to visit_assign?

That seems better than first moving furhter down the tree only to then peek back up using `location`.

> Should we ban Subtype projection depending on the MIR phase?

Yes. Currently this check should fail with `-Zvalidate-mir` since you are saying implicit subtyping is never allowed.